SU2_CFD/src/output/filewriter/CCGNSFileWriter.cpp

```diff
@@ -0,0 +1,256 @@
+  if (!dataSorter->GetConnectivitySorted()) {
+    SU2_MPI::Error("Connectivity must be sorted.", CURRENT_FUNCTION);
+  }
+
+  nLocalPoints = dataSorter->GetnPoints();
+  nDim = dataSorter->GetnDim();
+  GlobalElem = static_cast<cgsize_t>(dataSorter->GetnElemGlobal());
+  GlobalPoint = static_cast<cgsize_t>(dataSorter->GetnPointsGlobal());
+  cumulative = 0;
+
+  if (typeid(dataPrecision) == typeid(float))
+    dataType = RealSingle;
+  else if (typeid(dataPrecision) == typeid(double))
+    dataType = RealDouble;
+  else
+    SU2_MPI::Error("Wrong data type in CGNS output", CURRENT_FUNCTION);
+
+  /*--- If surface file cell dimension is decreased. ---*/
+  auto nCell = static_cast<int>(nDim - isSurface);
+
+  if (rank == MASTER_NODE) {
+    /*--- Remove the previous file if present. ---*/
+    remove(fileName.c_str());
+
+    /*--- Create CGNS file and open in write mode. ---*/
+    CallCGNS(cg_open(fileName.c_str(), CG_MODE_WRITE, &cgnsFileID));
+
+    /*--- Create Base. ---*/
+    CallCGNS(cg_base_write(cgnsFileID, "Base", nCell, nDim, &cgnsBase));
+
+    /*--- Create Zone. ---*/
+    array<cgsize_t, 3> zoneData;
+
+    zoneData[0] = GlobalPoint;
+    zoneData[1] = GlobalElem;
+    zoneData[2] = 0;
+
+    CallCGNS(cg_zone_write(cgnsFileID, cgnsBase, "Zone", zoneData.data(), Unstructured, &cgnsZone));
+  }
+}
+
+void CCGNSFileWriter::WriteField(int iField, const string& FieldName) {
+  /*--- Check if field is coordinate. ---*/
+  bool isCoord = false;
+  if (FieldName == "CoordinateX" || FieldName == "CoordinateY" || FieldName == "CoordinateZ") isCoord = true;
+
+  /*--- Create send buffer. ---*/
+  sendBufferField.resize(nLocalPoints);
+
+  for (unsigned long iPoint = 0; iPoint < nLocalPoints; iPoint++) {
+    sendBufferField[iPoint] = static_cast<dataPrecision>(dataSorter->GetData(iField, iPoint));
+  }
+
+#ifdef HAVE_MPI
+  if (rank != MASTER_NODE) {
+    SU2_MPI::Send(sendBufferField.data(), nLocalPoints * sizeof(dataPrecision), MPI_CHAR, MASTER_NODE, 0,
+                  SU2_MPI::GetComm());
+    return;
+  }
+#endif
+
+  /*--- Coordinate vector is written in blocks, one for each process. ---*/
+  cgsize_t nodeBegin = 1;
+  cgsize_t nodeEnd = static_cast<cgsize_t>(nLocalPoints);
+
+  if (isCoord) {
+    int CoordinateNumber = iField + 1;
+    CallCGNS(cg_coord_partial_write(cgnsFileID, cgnsBase, cgnsZone, dataType, FieldName.c_str(), &nodeBegin, &nodeEnd,
+                                    sendBufferField.data(), &CoordinateNumber));
+  } else {
+    int fieldNumber;
+    CallCGNS(cg_field_partial_write(cgnsFileID, cgnsBase, cgnsZone, cgnsFields, dataType, FieldName.c_str(), &nodeBegin,
+                                    &nodeEnd, sendBufferField.data(), &fieldNumber));
+  }
+
+#ifdef HAVE_MPI
+  for (int i = 0; i < size; ++i) {
+    if (i == MASTER_NODE) continue;
+    nodeBegin = static_cast<cgsize_t>(dataSorter->GetnPointCumulative(i) + 1);
+    nodeEnd = static_cast<cgsize_t>(dataSorter->GetnPointCumulative(i + 1));
```

Comment:
Add some comment here. It looks like CGNS is one-based and uses inclusive ranges? Is that correct?